R/tamper.R

```diff
@@ -0,0 +1,272 @@
+
+  repeat {
+
+    which <- menu(
+      c(other_text[pipe_stack + 1], stacks[[pipe_stack + 1]]),
+      title = titles[pipe_stack + 1])
+
+    if (which == 1) {
+      pipe_stack <- ! pipe_stack
+
+    } else if (which == 2) {
+      eval(substitute(browser(skipCalls = skip),
+                      list(skip = 7 - pipe_call)),
+           envir = sys.frame(pipe_call))
+
+    } else if (which > 1 && pipe_stack) {
+      which <- which - 2
+      eval(substitute(browser(skipCalls = skip),
+                      list(skip = 7 - freduce_calls[which])),
+           envir = sys.frame(freduce_calls[which]))
+
+    } else if (which > 1 && ! pipe_stack) {
+      which <- which - 1
+      eval(substitute(browser(skipCalls = skip),
+                      list(skip = 7-which)), envir = sys.frame(which))
+
+    } else {
+      break
+    }
+  }
+
+  ## ---------------------------------------------------------------
+
+}
+
+get_pipe_calls <- function(calls) {
+
+  freduce_calls <- integer()
+  for (i in seq_along(calls)) {
+    calli <- calls[[i]]
+    if (! is.name(calli[[1L]])) next
+    func <- as.character(calli[[1L]])
+    if (func == "freduce") {
+      freduce_calls <- c(freduce_calls, i)
+    }
+  }
+  freduce_calls
+}
+
+get_last_pipe_call <- function(calls) {
+
+  pipe_call <- length(calls) - 1L       # last one is tamper()
+  while (pipe_call >= 1) {
+    if (is.symbol(calls[[pipe_call]][[1]]) &&
+        as.character(calls[[pipe_call]][[1]]) == "%>%") break
+    pipe_call <- pipe_call - 1L
+  }
+  pipe_call
+}
+
+chain_parts_to_chr <- function(chain_parts) {
+  no_pipes <- length(chain_parts$pipes)
+  list(
+    lhs = paste(deparse(chain_parts$lhs), collapse = "\n    "),
+    pipes = vapply(chain_parts$pipes, as.character, ""),
+    rhss = vapply(lapply(chain_parts$rhss, deparse),
+      paste, "", collapse = "\n         ")
+  )
+}
+
+get_bad_stage <- function(freduce_calls) {
+
+  ## ---------------------------------------------------------------
+  ## We might be more down from the actual error in the pipe chain,
+  ## because of promises. For example if `value` is a promise all
+  ## along the chain, then we are at the last stage.
+  ##
+  ## We have no_pipes stages in total, and we have
+  ## length(freduce_calls) stages in the stack, these
+  ## correspond to the first length(freduce_calls) stages.
+  ##
+  ##        |--------|--------|--------|--------|--------|
+  ## pipes: |   1    |   2    |   3    |   4    |    5   |
+  ## stack: |   1    |   2    |   3    |
+  ##
+  ## Now we find which stage the error corresponds to. We need
+  ## to go up in the stack, until we find `value` evaluated.
+
+  ## Need to do this to allow debugging code that uses pryr itself
+  ## and has is_promise or promise_info
+  `_is_promise` <- pryr::is_promise
+  `_promise_info` <- pryr::promise_info
+
+  ## Checks which stacks are below the error in reality
+  ## If the `value` promise is not evaled we are below
+  is_below_error <- substitute(
```

Comment:
Can you supply an explicit 2nd argument here? Otherwise it's hard to see what's being substituted.